### PR TITLE
[EMCAL-792] Added checker for raw error check (all types, last minute)

### PR DESCRIPTION
**Pull request by anabananana98**
Added checker for raw error check (all types, last minute) which uses a moving average to determine if error rates are within threshold values

### CODE REVIEW

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+#include <TRobustEstimator.h>
+#include <ROOT/TSeq.hxx>
+#include <iostream>
+#include <vector>
+#include <TCanvas.h>
+#include <string>
+#include <TGraph.h>
+#include <queue>
+
+using namespace std;
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:21]: `Using` of whole namespaces must never be applied except the one declared in the header. In particular don't apply `using namespace std`, or of any member of the standard namespace.

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+void RawErrorCheckAll::configure()
+{
+    
+    // configure threshold-based checkers
+    auto nBadThreshold = mCustomParameters.find("BadThreshold");
+    if (nBadThreshold != mCustomParameters.end()) {
+        try {
+            mBadThreshold = std::stod(nBadThreshold->second);
+        } catch (std::exception& e) {
+            ILOG(Error, Support) << fmt::format("Value {} not a double", nBadThreshold->second.data()) << ENDM;
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:34]: For c++ streams like `ILOG` you don't need any formatting library, as this is handled by the iostream library. In any case, with c++20's std::format the format library anyhow kind of becomes obsolete.

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                    continue;
+                }
+                
+                // queue used to store list so that we get the average
+                queue<double> Dataset;
+                double sum;
+                double mean = -1;
+                
+                double *yValues = trendgraph->GetY();
+                double numPoints = trendgraph->GetN();
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:70]: You need int here. Better use `auto` for capturing return values.

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+
+
+Quality RawErrorCheckAll::check(std::map<std::string, std::shared_ptr<MonitorObject>>* moMap)
+{
+    Quality result = Quality::Good;
+    
+    for (auto& [moName, mo] : *moMap) {
+        if (mo->getName() == "TrendRawDataError") {
+            auto* c = dynamic_cast<TCanvas*>(mo->getObject());
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:56]: Meaningful names

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            TList* list_name = c->GetListOfPrimitives();
+            double counts = -1;
+            for (auto trendgraph : TRangeDynCast<TGraph>(list_name)) {
+                if (!trendgraph) {
+                    continue;
+                }
+                
+                // queue used to store list so that we get the average
+                queue<double> Dataset;
+                double sum;
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:66]: `sum` uninitialized, leads to random behaviour

**Review by mfasDa — CHANGES_REQUESTED** [review]
Still a few issues to fix

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+void RawErrorCheckAll::configure()
+{
+
+  // configure threshold-based checkers
+  auto nBadThreshold = mCustomParameters.find("BadThreshold");
+  if (nBadThreshold != mCustomParameters.end()) {
+    try {
+      mBadThreshold = std::stod(nBadThreshold->second);
+    } catch (std::exception& e) {
+      ILOG(Error, Support) << ("Value {} not a double", nBadThreshold->second.data()) << ENDM;
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:32]: Doesn't work - see other PR

File: Modules/EMCAL/src/RawErrorCheckAll.cxx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    }
+  }
+
+  auto nPeriodMovAvg = mCustomParameters.find("PeriodMovAvg");
+  if (nPeriodMovAvg != mCustomParameters.end()) {
+    try {
+      mPeriodMovAvg = std::stoi(nPeriodMovAvg->second);
+    } catch (std::exception& e) {
+      ILOG(Error, Support) << ("Value {} not an int", nPeriodMovAvg->second.data()) << ENDM;
```
Comment (mfasDa) [inline Modules/EMCAL/src/RawErrorCheckAll.cxx:41]: Same here

**Review by mfasDa — CHANGES_REQUESTED** [review]
Almost ready - please fix the two leftover cases and then I merge it.

**Review by mfasDa — APPROVED** [review]
Looks good now.

**Comment by knopers8** [comment]
please resolve the merge conflict.

**Pull request merged**